connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTask.java

```diff
@@ -458,13 +458,8 @@ boolean sendRecords() {
     }
 
     protected List<SourceRecord> poll() throws InterruptedException {
-        try {
-            return task.poll();
-        } catch (RetriableException | org.apache.kafka.common.errors.RetriableException e) {
-            log.warn("{} failed to poll records from SourceTask. Will retry operation.", this, e);
-            // Do nothing. Let the framework poll whenever it's ready.
-            return null;
-        }
+        retryWithToleranceOperator.reset();
+        return retryWithToleranceOperator.execute(task::poll, Stage.TASK_POLL, this.getClass());
```

Comment:
Furthermore, we aren't propagating any exceptions after retry exhaustion (due to timeout) so the task will never be marked as failed and we'll just keep calling `poll` in the execution loop for the worker source task. That definitely doesn't seem like the right thing to do as it defeats the purpose of configuring `errors.retry.timeout`, `errors.retry.delay.max.ms`, `errors.tolerance`.

Reply:
Yeah but I didn't think I really needed to handle that case and the existing methods should be taking care of it going by the example above with converters.

Reply:
Not sure why would the task not fail. As already explained [here](https://github.com/apache/kafka/pull/13726#discussion_r1218190316), if a non retriable exception is thrown, then 
based on this condition 


https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java#L225-L233

a ConnectException would be thrown or a null value is returned. It depends if the exception thrown from the specific operation is assignable or not and also on the tolerance limits. So, if the tolerance limit is none, then it fails. 

Also, if it's a retriableException and the deadline exceeds, then a null value is returned. The current code [here](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTask.java#L356) already handles null values returned from poll().